server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java

```diff
@@ -453,6 +453,10 @@ public boolean eagerGlobalOrdinals() {
         return false;
     }
 
+    public boolean fieldExists(SearchExecutionContext context) {
+        return context.getFieldInfos().fieldInfo(name) != null;
+    }
```

Comment:
Can you add javadocs to this method to explain that this API only exists to avoid failing due to the number of clauses on filebeat/metricbeat indices that have thousands of mapped fields but at most a couple hundreds that actually have data? And maybe a note that this is something that we hope to phase out in the future when most users are expected to no longer have these thousands of mapped fields without data?

Reply:
Also I'd suggest to make this method abstract to force implementations to think of the right thing to do. This would make the change larger, but it'd also be less trappy for fields like `constant_keyword`, `rank_feature` or `aggregate_metric_double` for which this implementation wouldn't be right? Someone creating a new field type in the future is also more likely to think of adding tests for `fieldsExists` if they have to implement it than if they inherit the default implementation.

Reply:
Or maybe alternatively change the default implementation to return `true` so that it would always be correct. And we would only have to correctly implement it on the most common field types like `text`, `keyword` and number fields to help users not hit exceptions about the number of clauses when querying their filebeat/metricbeat indices?

Reply:
Implementing this only on mapping types that could be created via dynamic mappings makes sense to me.  There are definitely some fields that we should implement this on to always return `false` though - `rank_feature` being a good example, I don't think that should ever be included in a general search across all fields, so maybe making it abstract and forcing us to consider each field type individually is a good idea.  I'll try it out.

I don't know that we plan to phase it out though? I think it's generally useful - I imagine that users quite often have mappings that have accreted over the years with some fields dropping out of use, or use shared mappings for multiple different data types and so have some fields that are never used.

Reply:
> Implementing this only on mapping types that could be created via dynamic mappings makes sense to me.

Note that this was not my suggestion. At least when a field is created via a dynamic mapping update, there is a guarantee that the field will have data: the document that triggered the dynamic mapping update. The case we are seeing with filebeat/metricbeat indices is different: index templates have thousands of concrete fields that never get populated with any data.

> There are definitely some fields that we should implement this on to always return false though - rank_feature being a good example, I don't think that should ever be included in a general search across all fields, so maybe making it abstract and forcing us to consider each field type individually is a good idea.

Interesting... and we'd rename the API to something like `boolean expandWildcards()` to better reflect its purpose?

I don't know if it's possible but another approach I had thought of would consist of making `MappedFieldType#termQuery` produce a `MatchNoDocsQuery` when the field doesn't exist in the index and intercepting places where we add clauses to the boolean query in order to ignore `MatchNoDocsQuery` as `SHOULD` clauses.

> I don't know that we plan to phase it out though? I think it's generally useful - I imagine that users quite often have mappings that have accreted over the years with some fields dropping out of use, or use shared mappings for multiple different data types and so have some fields that are never used.

My reasoning is that the instance of this problem I'm the most familiar with are filebeat/metricbeat indices, and that this problem is resolved via Fleet, which creates separate data streams where each data stream only has the relevant fields. Unused fields in mappings not only creates problems with numbers of clauses in boolean queries, it also makes the cluster state larger than it should be and it makes Kibana return poor field suggestions. I had thought of this change as a tactical change to better help users of legacy filebeat/metricbeat indices migrate as we remove the ability to configure the number of clauses. If we want to make it more generally useful, we might need to think more of how to handle e.g. field suggestions.

For the record, fields dropping out of use would not be covered by checks on FieldInfos since Lucene's FieldInfos are append-only. In the case when all documents that have a given field get deleted and merged away, the field would still exist in FieldInfos but would return empty Terms/PointValues/DocValues instances.

Reply:
> I don't know if it's possible but another approach I had thought of would consist of making MappedFieldType#termQuery produce a MatchNoDocsQuery when the field doesn't exist in the index and intercepting places where we add clauses to the boolean query in order to ignore MatchNoDocsQuery as SHOULD clauses.

I wanted to try this originally, but unfortunately the query parsers separate expanding wildcard fields and generating queries, and the field limit checks happen before we call `termQuery` on anything.